### Experimento numérico \# 3- Tema: Variación de tamaños de matrices

**Responsable:** Danahi Ramos Martínez

## 0. Objetivo

Este documento pretende mostar los resultados obtenidos con la implementación realizada, en el contexto del comportamiento del método de solución de bloques al variar tamaño de matriz de sistema.

**Nota:** El reporte general de resultados obtenidos con las implementaciones realizadas en este proyecto se puede consultar en el siguiente [vínculo](https://github.com/mno-2020-gh-classroom/ex-modulo-3-comp-matricial-svd-czammar/blob/master/results/Reporte_resultados.ipynb)

## 1. Consideraciones

Al respecto de los experimentos numéricos realizados para consolidar el presente reporte, tales pretenden mostrar el comportamiento del método de solución de bloques al variar tamaño de matriz de sistema

En este sentido, en particular para cada experimento realizado, se reportan según resulte aplicable:

* los parámetros empleados en la simulaciones, 
* las dimensiones de las matrices y vectores involucrados, así como el prodecimiento pseudo-aleatorio que les dio origen, 
* 1) el tiempo involucrado en correr los experimentos, 
* 2) número de condición de las matrices pseudo-aleatorias, y 
* 3) el error relativo obtenido, para la solución de un sistema lineal de la forma $Ax=b$, es decir el cociente

$$|| Ax - b ||_2 / || b ||_2$$
* 4) La norma de $A$ y de la matriz obtenida al rearmar A tras multiplica las matrices resultantes de la aproximación de la descomposición SVD vía el algoritmo **One-Sided Jacobi**


### 1.1 Consideraciones sobre la infraestructura empleada

**Especificaciones de ambiente común de trabajo**

Para realizar el presente experimento numérico se ha empleado la imagen de docker basada en R del curso MNO 2020 (palmoreck/jupyterlab_r_kernel:1.1.0)

```
docker run --rm -v `pwd`:/datos --name jupyterlab_r_kernel_local -p 8888:8888 -d palmoreck/jupyterlab_r_kernel:1.1.0

```

*Nota:* el comando "-v \`pwd\`:/datos", permite montar el directorio actual en donde el usuario se encuentre situada en terminal como un volumen de la imagen de docker, dentro del directorio "/datos".

A mayor abundamiento, el siguiente experimento fue realizado en una máquina con las siguientes características:
MacBook Pro (13-inch, 2019, Four Thunderbolt 3 ports)
* Procesador: 2.4 GHz Intel Core i5 de cuatro núcleos
* Memoria: 8 GB 2133 MHz LPDDR3


## 2. Experimento numérico

**Objetivo**

Probar el desempeño de la eliminación por bloques por medio de SVD con respecto a diferentes tamaños de matrices, es decir matrices de tamaño: 10x10 y 100x100 

A tal respecto, se destaca que se realizó este experimentos buscando probar [Pendiente: desarrollo]

**Cargamos codigo desarrollado previamente**

* **utils.R:** contiene las funciones auxiliares desarrolladas para el proyecto
* **00-load.R:** la implementación del método de eliminación por bloques, empleando la aproximación de la descomposición SVD vía el algoritmo **One-Sided Jacobi**.

In [2]:
## Instalamos paquetes
rm(list = ls())

paquetes <- c('matrixcalc')

instalar <- function(paquete) {
  if (!require(paquete,character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)) {
    install.packages(as.character(paquete), dependecies = TRUE, repos = "http://cran.us.r-project.org")
    library(paquete, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)
  }
}


lapply(paquetes, instalar)

[[1]]
NULL

In [3]:
## Cargamos paquetes necesarios
library("matrixcalc")
library(matlib)
library(tictoc)
#source("metadata.R")
source("utils.R")
source("00-load.R")

Warning message in rgl.init(initValue, onlyNULL):
“RGL: unable to open X11 display”
Warning message:
“'rgl.init' failed, running with 'rgl.useNULL = TRUE'.”

Attaching package: ‘matlib’


The following object is masked from ‘package:matrixcalc’:

    vec




### 2.1 Experimento 3.1 

Este experimento se basa en los siguientes parámetros [Pendiente: por favor ajustar tabla según experimento]

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     |     10x10              |             |
| 2 |        Dimensión $b$       |     10x1               |             |
| 3 |         Tolerancia         |     $10^{-8}$            |             |
| 4 |          Maxsweep          |     5                  |             |
| 5 |          corte             |     5                  | Corte de bloque            |
| 6 |   Metodo pseudo-aleatorio  |     números reales     |Números aleatorios de la distribución normal estándar|
| 7 | Numero de condición de $A$ |     9.74               |cond(A) = $|| A ||.|| A^{-1}||$     |


**Codigo:** Fijando una semilla y creando una matriz de tamaño 10x10 por medio de un método pseudoaleatorio conforme a la distribución normal estandandarizada, se probo el desempeño de la función eliminación por bloques.

In [19]:
set.seed(231)
n= 10**1

A = matrix(rnorm(n**2), ncol=n)
b = matrix(rnorm(n), ncol=1)
TOL = 10**-8
tic()
z<-eliminacion_bloques(A,b,n/2,10^-8,5)
toc()


#Condición
norm(A,"2")*norm(inv(A),"2")
#Error relativo
norm(A%*%z-b,"2")/norm(b,"2")
#Error factorización
svd_aprox = svd_jacobi_aprox(A,TOL,5)
norm(A-svd_aprox$U%*%svd_aprox$S%*%t(svd_aprox$V),"2")

0.059 sec elapsed


[1] 9.749164

[1] 1.787496

[1] 3.670281e-15

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla [Pendiente: por favor ajustar tabla según experimento]

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |      0.059 sec           |                                   |
| 2 |      Error relativo $||Ax-b||/||b||$     |      1.78749566520204  |                                   |
| 3 | Error factorizacion SVD $||A-U\cdot \Sigma \cdot V^T||$ |  3.67028106314907e-15 |                     |


### 2.2  Experimento 3.2

Este experimento se basa en los siguientes parámetros 

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     |    100x100                    |             |
| 2 |        Dimensión $b$       |    100x1                    |             |
| 3 |         Tolerancia         |    $10^{-8}$                     |             |
| 4 |          Maxsweep          |    5                    |             |
| 5 |     Corte    |    50                  |             |
| 6 |   Metodo pseudo-aleatorio  |     números reales     |Números aleatorios de la distribución normal estándar|
| 7 | Numero de condición de $A$ |     109.17              |cond(A) = $|| A ||.|| A^{-1}||$     |

**Codigo:** Fijando una semilla y creando una matriz de tamaño 100x100 por medio de un método pseudoaleatorio conforme a la distribución normal estandandarizada, se probo el desempeño de la función eliminación por bloques.

In [20]:
set.seed(231)
n= 10**2

A = matrix(rnorm(n**2), ncol=n)
b = matrix(rnorm(n), ncol=1)
TOL = 10**-8
tic()
z<-eliminacion_bloques(A,b,n/2,10^-8,5)
toc()


#Condición
norm(A,"2")*norm(inv(A),"2")
#Error relativo
norm(A%*%z-b,"2")/norm(b,"2")
#Error factorización
svd_aprox = svd_jacobi_aprox(A,TOL,5)
norm(A-svd_aprox$U%*%svd_aprox$S%*%t(svd_aprox$V),"2")

37.399 sec elapsed


[1] 109.1753

[1] 1.163344

[1] 9.094827e-14

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla:

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |      37.399 sec          |                                   |
| 2 |      Error relativo $||Ax-b||/||b||$     |      1.1633441561769 |                                   |
| 3 | Error factorizacion SVD $||A-U\cdot \Sigma \cdot V^T||$ |  9.09482723558874e-14 |                     |

**Resumen de resultados obtenidos en experimentos**

Los resultados obtenidos, se compararan a través de la siguiente:

| # | Experimento | Tiempo de desempeño | Error relativo | Error Factorización |
|:-:|:-----------:|:-------------------:|----------------|----------------|
| 1 |     3.1     | 0.059 sec           |1.78749566520204|3.67028106314907e-15 |    
| 2 |     3.2     | 37.399              |1.1633441561769 |9.09482723558874e-14                | 



## 3. Principales hallazgos

Al respecto, se destacan los siguientes hallazgos del experimento:

* Hallazgo 1: El tiempo aumenta considerablemente conforme aumenta el tamaño de la matriz
* Hallazgo 2: El error relativo disminuye aumentando el tamaño de matriz